jobs/audit_jobs.py

```diff
@@ -89,13 +99,50 @@ def run(self):
 
         model_groups = python_utils.ZIP(KIND_BY_INDEX, models_of_kind_by_index)
         for kind, models_of_kind in model_groups:
-            # NOTE: Using extend() instead of append() because ApplyAuditDoFns
-            # produces an iterable of PCollections rather than a single one.
-            # NOTE: Label is missing because ApplyAuditDoFns labels itself.
             audit_error_pcolls.extend(models_of_kind | ApplyAuditDoFns(kind))
 
+            if kind in KINDS_REFERENCED_BY_PROPS:
+                existing_key_pcolls.append(
+                    models_of_kind | GetExistingModelKeys(kind))
+
+            if kind in PROP_REFERENCES_BY_KIND:
+                errors_by_expected_key_pcolls.extend(
+                    models_of_kind | GetErrorsByExpectedModelKey(kind))
+
+        counts_by_existing_key = (
+            existing_key_pcolls
+            | 'Flatten PCollections of actual keys' >> beam.Flatten()
+        )
+        errors_by_expected_key = (
+            errors_by_expected_key_pcolls
+            | 'Flatten PCollections of expected keys' >> beam.Flatten()
+        )
+        audit_error_pcolls.append(
+            (counts_by_existing_key, errors_by_expected_key)
+            | 'Group counts and errors by key' >> beam.CoGroupByKey()
+            | 'Filter keys without any errors' >> beam.FlatMapTuple(
+                self._get_model_relationship_errors)
+        )
+
         return audit_error_pcolls | 'Combine audit results' >> beam.Flatten()
 
+    def _get_model_relationship_errors(self, unused_key, counts_and_errors):
```

Comment:
> I don't understand why we're passing "key" in if it's unused...

Because that's the return value of CoGroupByKey()

> Also the docstring talks about doing something with the "given model key". But in that case why are we ignoring it?
>
> EDIT: Ah, I see, the impl does not depend on the details in the key. Maybe explain that somewhere, otherwise it looks a bit like an error.

Right, the job of the keys have already been fulfulled in the CoGroupByKey stage, that's why it's safe to ignore it. Maybe `unused_grouped_by_key`?

Reply:
Hm, that's a bit hard to parse. I lean towards just keeping the existing name and adding some documentation to explain.